integration/benchmark/cmd/ci/main.go

```diff
@@ -271,3 +298,96 @@ func sendDataToBigQuery(
 	}
 	return nil
 }
+
+// adjustTPS tries to find the maximum TPS that the network can handle using a simple AIMD algorithm.
+// The algorithm starts with minTPS as a target.  Each time it is able to reach the target TPS, it
+// increases the target by `additiveIncrease`. Each time it fails to reach the target TPS, it decreases
+// the target by `multiplicativeDecrease` factor.
+//
+// To avoid oscillation and speedup conversion we skip the adjustment stage if TPS grew
+// compared to the last round.
+//
+// Target TPS is always bounded by [minTPS, maxTPS].
+func adjustTPS(
+	lg *benchmark.ContLoadGenerator,
+	log zerolog.Logger,
+	interval time.Duration,
+	minTPS, maxTPS, maxInflight uint,
+) error {
+	targetTPS := minTPS
+	lastTs := time.Now()
+	lastTPS := float64(minTPS)
+	lastTxs := uint(lg.GetTxExecuted())
+	for {
+		select {
+		case <-time.After(interval):
+			currentSentTxs := lg.GetTxSent()
+			currentTxs := uint(lg.GetTxExecuted())
+			inflight := currentSentTxs - int(currentTxs)
+			currentTPS := float64(currentTxs-lastTxs) / math.Floor(time.Since(lastTs).Seconds())
+
+			// Do not touch target TPS if TPS rate incresed since last check.
+			if (currentTPS > float64(lastTPS)) && (inflight < int(maxInflight)) {
+				log.Info().
+					Float64("lastTPS", lastTPS).
+					Float64("currentTPS", currentTPS).
+					Int("inflight", inflight).
+					Msg("skipped adjusting TPS")
+
+				lastTxs = currentTxs
+				lastTPS = currentTPS
+				lastTs = time.Now()
```

Comment:
you're using different "now" time within the loop (here, time.Since, etc).  maybe just use the one returned by `<-time.After(interval)`